FernFlower-Patches/0035-Make-decomp-threaded.patch

```diff
@@ -0,0 +1,804 @@
+ import java.util.ArrayList;
+ import java.util.List;
++import java.util.concurrent.ExecutionException;
++import java.util.concurrent.ExecutorService;
++import java.util.concurrent.Executors;
++import java.util.concurrent.Future;
+ import java.util.jar.JarFile;
+ import java.util.jar.Manifest;
++import java.util.stream.Collectors;
++import java.util.stream.IntStream;
+ import java.util.zip.ZipFile;
+ 
+ public class ContextUnit {
+@@ -118,7 +124,10 @@ public class ContextUnit {
+           StructClass cl = classes.get(i);
+           String entryName = decompiledData.getClassEntryName(cl, classEntries.get(i));
+           if (entryName != null) {
+-            String content = decompiledData.getClassContent(cl);
++            String content = null;
++            if (decompiledData.processClass(cl)) {
++              content = decompiledData.getClassContent(cl);
++            }
+             if (content != null) {
+               int[] mapping = null;
+               if (DecompilerContext.getOption(IFernflowerPreferences.BYTECODE_SOURCE_MAPPING)) {
+@@ -149,13 +158,64 @@ public class ContextUnit {
+           }
+         }
+ 
+-        // classes
+-        for (int i = 0; i < classes.size(); i++) {
+-          StructClass cl = classes.get(i);
+-          String entryName = decompiledData.getClassEntryName(cl, classEntries.get(i));
+-          if (entryName != null) {
+-            String content = decompiledData.getClassContent(cl);
+-            resultSaver.saveClassEntry(archivePath, filename, cl.qualifiedName, entryName, content);
++        //Whooo threads!
++        int threads = DecompilerContext.getThreads();
++        if (threads > 1) {
++
++          DecompilerContext rootContext = DecompilerContext.getCurrentContext();
++          ExecutorService executor = Executors.newFixedThreadPool(threads);
++
++          //Compute the classes we need to decomp.
++          List<ClassContext> toProcess = IntStream.range(0, classes.size()).parallel()
++            .mapToObj(i -> {
++              StructClass cl = classes.get(i);
++              return new ClassContext(cl, decompiledData.getClassEntryName(cl, classEntries.get(i)));
++            })
++            .filter(e -> e.entryName != null)
++            .collect(Collectors.toList());
++          List<Future<?>> futures = new ArrayList<>(toProcess.size());
++
++          //Submit preprocessor jobs.
++          for (ClassContext clCtx : toProcess) {
++            futures.add(executor.submit(() -> {
++              DecompilerContext.cloneContext(rootContext);
++              clCtx.ctx = DecompilerContext.getCurrentContext();
++              clCtx.shouldContinue = decompiledData.processClass(clCtx.cl);
++              DecompilerContext.setCurrentContext(null);
++            }));
++          }
++
++          //Ask the executor to shutdown
++          executor.shutdown();
++          waitForAll(futures);
++          futures.clear();
++
++          executor = Executors.newFixedThreadPool(threads);
```

Comment:
Why shutdown the executor just to recreate another of the same size?

Reply:
Shutting down the Executor ensures it actually finishes processing all the submitted tasks and we don't just sit there waiting for nothing.  Unless your sure that's redundant to do.

Reply:
From the javadoc on `ExecutorService#shutdown`:
```
Initiates an orderly shutdown in which previously submitted
tasks are executed, but no new tasks will be accepted.
Invocation has no additional effect if already shut down.
```
With a fixed thread pool all the threads will block until there is work to be done in the queue, so the tasks will be completed as soon as there's a free thread to do them regardless of whether the pool itself is shutdown.

Reply:
I've had issues with things like this in the past, and iirc, the fix was to ask the executor to shutdown and wait for all the Futures to finish. I'd rather leave it the way it is and not have to re-run days of burnin tets. It's not that big a deal to throw away the executor, there's other performance issues in FF besides this.